### PR TITLE
Created login/user registration interface

### LINKED ISSUE #36: Setup login feature

### PR DESCRIPTION
**Pull request by jeffreyHoelzel**
Finally got everything working and should be ready to be deployed, I am very tired so apologies if any code or comments make no sense. But a person can create a simple account with their name, email, username, and password, then log in with their username and password as you would expect. Please provide all the feedback you can, the styling and stuff like the timeout after pressing a button can be modified as we decide.

### CODE REVIEW

**Comment by jeffreyHoelzel** [comment]
So I went in and resolved all the merge conflicts and then realized that a person reviewing has to do that, if anyone wants to hop on a call and go through each conflict, I can help with resolving them.

File: Instructions.txt
Comment (SeanGolez) [inline Instructions.txt:1]: Remove this because it has been changed.

File: db_model.mwb
Comment (SeanGolez) [inline db_model.mwb:1]: I created a model directory in main where this stuff is kept

File: db_model.mwb.bak
Comment (SeanGolez) [inline db_model.mwb.bak:1]: I created a model directory in main where this stuff is kept

File: service/Dockerfile
Comment (SeanGolez) [inline service/Dockerfile:1]: That top line is commented out in main

File: service/app.py [hunk 1]
```diff
@@ -23,24 +52,76 @@ def send_message():
...
+        password = data.get('password')
+        email = data.get('email')
+        first_name = data.get('firstName')
+        last_name = data.get('lastName')
+
+        logger.info('\nusername: %s', username)
+        logger.info('\npassword: %s', password)
+        logger.info('\nemail: %s', email)
+        logger.info('\nfirst name: %s', first_name)
+        logger.info('\nlast name: %s', last_name)
```
Comment (SeanGolez) [inline service/app.py]: these logs for debugging are not needed

File: service/app.py [hunk 2]
```diff
@@ -23,24 +52,76 @@ def send_message():
...
+
+        # if all credentials are not empty strings, create a new user object, otherwise, throw error
+        if username != '' and password != '' and email != '' and first_name != '' and last_name != '':
+            new_user = User(username=username, password=password, email=email, first_name=first_name, last_name=last_name)
```
Comment (SeanGolez) [inline service/app.py:77]: I think that something like:
```
if all([username, password, email, first_name, last_name]):
```
would be more concise and checks for the same thing, but you could argue that is less explicit

File: web/src/DBExample.js
Comment (SeanGolez) [inline web/src/DBExample.js:1]: this is not in main anymore

File: web/src/components/AuthenticateUser.js
Comment (SeanGolez) [inline web/src/components/AuthenticateUser.js:1]: I'm not really familiar with using localStorage to save info, but this seems good!

File: web/src/components/AuthenticateUser.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    // check if login status is true
+    if (storedLoginStatus === "true") {
+      // set status to true as bool and username
+      setLoginStatus(true);
+      setUsername(storedUsername);
+    } else {
+      // set status to false as bool and empty string to username
+      setLoginStatus(false);
+      setUsername("");
```
Comment (SeanGolez) [inline web/src/components/AuthenticateUser.js]: I'm not sure if this else is needed if the variables are initialized with those values.

File: web/src/components/Login.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        setMessage(data.message || "Error logging in.");
+      }
+
+      // set message and validation to confirm or deny login
+      setMessage(data.message);
+      setValid(data.exists);
+
+    } catch (error) {
+      console.log("Error:", error);
+      setMessage("Trouble connecting to backend.");
```
Comment (SeanGolez) [inline web/src/components/Login.js]: Maybe make this "Error logging in" too since the user shouldn't know about how it works.

**Review by SeanGolez — CHANGES_REQUESTED** [review]
I think this looks really great!
I recommend that you merge the main branch into your branch first to fix all of the merge conflicts and inconsistencies with your pull request.
Also, you should remove the logging lines in the backend because those will not be there in the final product.

File: web/src/components/RegisterUser.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        setMessage(data.message || "Error logging in.");
+      }
+
+      // set message and validation to confirm or deny login
+      setMessage(data.message);
+      setError(data.error);
+
+    } catch (error) {
+      console.log("Error:", error);
+      setMessage("Trouble connecting to backend.");
```
Comment (SeanGolez) [inline web/src/components/RegisterUser.js]: Same thing about making this "Error logging in"

File: web/src/components/Header.js [hunk 6]
```diff
@@ -1,9 +1,34 @@
...
 const Header = () => {
+    // set up use state for displaying login popup
+    const [loginPopup, setLoginPopup] = useState(false);
+    // set up authentication hook
+    const {loginStatus, username} = AuthenticateUser();
+
+    // allow login popup to be toggleable
+    const toggleLoginPopup = () => {
+      setLoginPopup(!loginPopup);
+    //   localStorage.clear(); // FOR TESTING
```
Comment (LukeBowen85) [inline web/src/components/Header.js]: this line can prob be removed if it was just used for testing

**Review by LukeBowen85 — CHANGES_REQUESTED** [review]
looks good overall aside from the couple things Sean mentioned too, think the only big thing was making sure to merge main into your branch to take care of any conflicts and it should be all good

File: service/app.py (hunk 2 above)
Comment (jeffreyHoelzel) [reply service/app.py:77]: I think it is more concise, added it.

File: web/src/components/AuthenticateUser.js (hunk 3 above)
Comment (jeffreyHoelzel) [reply web/src/components/AuthenticateUser.js]: Agreed

File: web/src/components/Login.js (hunk 4 above)
Comment (jeffreyHoelzel) [reply web/src/components/Login.js]: Agreed

**Pull request merged**